make_prg/utils/io_utils.py

```diff
@@ -0,0 +1,219 @@
+            if file_should_be_cleared:
+                filename = Path(filename).with_suffix("").name
+            else:
+                return filename
+
+    @staticmethod
+    def _delete_file(filepath: Optional[Path]):
+        if filepath is not None:
+            filepath.unlink(missing_ok=True)
+
+    def delete_files(self):
+        self._delete_file(self.PRG)
+        self._delete_file(self.binary_PRG)
+        self._delete_file(self.gfa)
+        self._delete_file(self.pickle)
+        self._delete_file(self.stats)
+
+    @staticmethod
+    def get_locus_to_set_of_output_files(temp_root: Path) -> Dict[str, "SetOutputFiles"]:
+        locus_to_set_of_output_files = defaultdict(SetOutputFiles)
+        for file in temp_root.iterdir():
+            if file.is_file():
+                locus_name = SetOutputFiles.clear_temp_extensions(file.name)
+                if file.name.endswith(".prg.fa"):
+                    locus_to_set_of_output_files[locus_name].PRG = file
+                elif file.name.endswith(".bin"):
+                    locus_to_set_of_output_files[locus_name].binary_PRG = file
+                elif file.name.endswith(".gfa"):
+                    locus_to_set_of_output_files[locus_name].gfa = file
+                elif file.name.endswith(".pickle"):
+                    locus_to_set_of_output_files[locus_name].pickle = file
+                elif file.name.endswith(".stats"):
+                    locus_to_set_of_output_files[locus_name].stats = file
+        return locus_to_set_of_output_files
+
+
+# Note: not unit tested
+def get_stats_on_variants(stats_files: List[Path]) -> Tuple[int, int]:
+    nb_of_variants_successfully_applied = 0
+    nb_of_variants_that_failed_to_be_applied = 0
+    for stat_file in stats_files:
+        with open(stat_file) as stat_file_fh:
+            line_split = stat_file_fh.readline().strip().split()
+            nb_of_variants_successfully_applied_for_this_locus = int(line_split[1])
+            nb_of_variants_successfully_applied += (
+                nb_of_variants_successfully_applied_for_this_locus
+            )
+            nb_of_variants_that_failed_to_be_applied_for_this_locus = int(line_split[2])
+            nb_of_variants_that_failed_to_be_applied += (
+                nb_of_variants_that_failed_to_be_applied_for_this_locus
+            )
+    return nb_of_variants_successfully_applied, nb_of_variants_that_failed_to_be_applied
+
+
+# this avoids circular import issues, probably has a better way through this though
+from make_prg import prg_builder
+
+
+# Note: not unit tested
+def create_final_files(temp_dir: Path, output_prefix: str, is_a_single_MSA: bool, output_stats: bool = False):
+    logger.info("Concatenating files from several threads into single final files...")
+
+    logger.info("Creating FASTA file of PRGs...")
+    locus_to_set_of_output_files = SetOutputFiles.get_locus_to_set_of_output_files(temp_dir / "mp_temp")
```

Comment:
> so am I right that your new code allows constructing multiple loci in parallel right?

yes

> and then, it concatenates them?

yes

> Does it also make separate PRG files? I would need that for gramtools so that i can then add in the invariant bits of the genome in between
> specifically- i need the .bin files to be kept separate
> i think that's what you do in lines 185-190 below- sorry if i'm making you repeat yourself

Don't worry about it, first of all, I am making a bit messy for you to review this PR, as I made it so large that I had to break into modules, but modules depend on one another... so you have to review the PR with incomplete information, which is far from ideal. Although this code does not make separate PRGs (a single fasta with all PRG strings), we do build separate binary representation of PRGs, that can be used by `gramtools`. As a concrete example, this is the output of an integration test not yet PRed here: https://github.com/leoisl/make_prg/tree/update/tests/integration_tests/data/truth_output/several . If you download `several.prg.bin.zip`, you will see the separate `.bin` files. I think this means all is good for you and `gramtools`? Could you please confirm?

Reply:
perfect